### PR TITLE
FEATURE: add Azure's OpenAI support

### PR DESCRIPTION
**Pull request by fokx**
* add Azure's OpenAI support
    * add custom endpoint URL in settings, in a similar way as how `discourse-ai` use Azure
    * Azure-related configurations done with `OpenAI.configure` in `bots/open_ai_bot.rb` `initialize`
* update `ruby-openai` to latest version
    * add its dependencies `faraday` and remove `httparty`
    * change response.parsed_response["key"] to just response["key"] (see `ruby-openai`'s [CHNAGELOG](https://github.com/alexrudall/ruby-openai/blob/main/CHANGELOG.md))

### CODE REVIEW

**Comment by fokx** [comment]
The Azure OpenAI application and deployment can be a little cumbersome, but I do managed to get the Azure version of gpt3.5-turbo-16k chatbot work on my forum.

**Comment by merefield** [comment]
I've merged an update which now uses 4.2.0 so some of these changes are redundant ... please merge in main.

**Comment by fokx** [comment]
OK, I have merged branch your 'main' into my 'support-azure-openai' branch.

**Review by merefield — CHANGES_REQUESTED** [review]

File: plugin.rb [hunk 1]
```diff
@@ -5,8 +5,12 @@
 # authors: merefield
 # url: https://github.com/merefield/discourse-chatbot
 
-gem "httparty", '0.21.0'
-gem "ruby-openai", '3.7.0', { require: false }
+gem 'faraday-net_http', '3.0.2', {require: false }
+gem 'ruby2_keywords', '0.0.5', {require: false }
+gem 'faraday', '2.7.10', {require: false }
```
[the remark this answers is not in the document]
Comment (merefield) [reply plugin.rb]: this appears to have been addressed, resolving.

File: config/locales/server.en.yml [hunk 2]
```diff
@@ -1,7 +1,10 @@
 en:
   site_settings:
     chatbot_enabled: "Enable the chatbot plugin"
-    chatbot_open_ai_token: "Your Open AI token. You can get one at <a target='_blank' rel='noopener' href='https://platform.openai.com/account/api-keys/'>openai.com</a>"
+    chatbot_openai_gpt4_url: "Modify it if you want to use GPT 4 by Azure, e.g. https://custom-domain.openai.azure.com/openai/deployments/DEPLOYMENT)."
```
[the remark this answers is not in the document]
Comment (fokx) [reply config/locales/server.en.yml:4]: When deploying GPT 3.5 / GPT 4 model in Azure's AI Studio, the `<CUSTOM_DEPLOYMENT>` part in the URL is also customizable.
So I think it is necessary to separate these to two different setting names.

PS: Azure OpenAI's URL is `https://<CUSTOM_SUBDOMAIN>.openai.azure.com/openai/deployments/<CUSTOM_DEPLOYMENT>`

File: config/settings.yml [hunk 3]
```diff
@@ -112,3 +115,5 @@ plugins:
   chatbot_enable_verbose_console_response_progress_logging:
     client: false
     default: false
+  chatbot_openai_gpt35_url: "https://api.openai.com/v1/chat/completions"
+  chatbot_openai_gpt4_url: "https://api.openai.com/v1/chat/completions"
```
[the remark this answers is not in the document]
Comment (fokx) [reply config/settings.yml:128]: Yeah, it is truly a duplication for OpenAI's API, but not for Azure's.

File: plugin.rb [hunk 4]
```diff
@@ -32,9 +36,9 @@ def progress_debug_message(message)
 
 after_initialize do
   SeedFu.fixture_paths << Rails
-    .root
-    .join("plugins", "discourse-chatbot", "db", "fixtures")
-    .to_s
+                            .root
```
[the remark this answers is not in the document]
Comment (fokx) [reply plugin.rb:38]: OK, I will use rubocop. I previously used RubyMine's `Alt+Shift+L` reformating.

File: config/locales/server.en.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (merefield) [reply config/locales/server.en.yml:4]: I think we could expect the user to change the URI setting if they change model, no?  The danger is other models will be released and are you planning on having even more settings to accommodate each and every released model?  I would be much more comfortable here with one setting, e.g. `chatbot_openai_azure_url` and require the user to update that as well as the model choice if required.

File: config/settings.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (merefield) [reply config/settings.yml:128]: I still don't understand why we need a definition per model?  If it changes per model then we should require the administrator to update the URL at the same time.

File: config/locales/server.en.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fokx) [reply config/locales/server.en.yml:4]: Yeah, I agree. One place for setting the URL of the current model used is enough. Admins can record their deployment URLs elsewhere anyway.

File: config/settings.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (fokx) [reply config/settings.yml:128]: I agree with you.

File: config/locales/server.en.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (merefield) [reply config/locales/server.en.yml]: thanks @fokx.  One more change please: could we make the setting name just slightly more consistent: so like `chatbot_azure_open_ai_token` can we have `chatbot_azure_open_ai_model_url` ... then I think we are all but done here!  thx

[the remark this answers is not in the document]
Comment (fokx) [reply config/locales/server.en.yml]: OK, I have renamed that option.

**Review by merefield — APPROVED** [review]
Thanks @fokx 🙏 I've made some minor changes to facilitate the merge.  Good work!

**Pull request merged**